swabbie-orca/src/main/kotlin/com/netflix/spinnaker/swabbie/orca/OrcaTaskMonitoringAgent.kt

```diff
@@ -0,0 +1,214 @@
+import java.time.Instant
+import java.time.temporal.Temporal
+import java.util.*
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+import javax.annotation.PostConstruct
+import javax.annotation.PreDestroy
+
+@Component
+class OrcaTaskMonitoringAgent (
+  private val clock: Clock,
+  val registry: Registry,
+  private val swabbieProperties: SwabbieProperties,
+  private val taskTrackingRepository: TaskTrackingRepository,
+  private val orcaService: OrcaService,
+  private val applicationEventPublisher: ApplicationEventPublisher,
+  private val retrySupport: RetrySupport,
+  private val lockingService: Optional<LockingService>
+) : DiscoveryActivated, MetricsSupport(registry) {
+
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+  private val worker: Scheduler.Worker = Schedulers.io().createWorker()
+
+  private val timeoutMillis: Long = 5000
+  private val maxAttempts: Int = 3
+
+  override val onDiscoveryUpCallback: (event: RemoteStatusChangedEvent) -> Unit
+    get() = {
+      worker.schedulePeriodically({
+       withLocking(javaClass.simpleName) { monitorOrcaTasks() }
+      }, getAgentDelay(), getAgentFrequency(), TimeUnit.SECONDS)
+    }
+
+  override val onDiscoveryDownCallback: (event: RemoteStatusChangedEvent) -> Unit
+    get() = { stop() }
+
+  @PostConstruct
+  private fun init() {
+    log.info("Initializing agent ${javaClass.simpleName}")
+    registry.gauge(lastRunAgeId.withTag("agentName", javaClass.simpleName), this) {
+      Duration
+        .between(it.getLastAgentRun(), clock.instant())
+        .toMillis().toDouble()
+    }
+  }
+
+  @PreDestroy
+  private fun stop() {
+    log.info("Stopping agent ${javaClass.simpleName}")
+    if (!worker.isUnsubscribed) {
+      worker.unsubscribe()
+    }
+  }
+
+  private fun withLocking(
+    agentName: String,
+    callback: () -> Unit
+  ) {
+    if (lockingService.isPresent) {
+      val normalizedLockName = (agentName)
+        .replace(":", ".")
+        .toLowerCase()
+      val lockOptions = LockManager.LockOptions()
+        .withLockName(normalizedLockName)
+        .withMaximumLockDuration(lockingService.get().swabbieMaxLockDuration)
+      lockingService.get().acquireLock(lockOptions) {
+        callback.invoke()
+      }
+    } else {
+      log.warn("***Locking not ENABLED***")
+      callback.invoke()
+    }
+  }
+
+  private fun monitorOrcaTasks() {
+    initialize()
+
+    val inProgressTasks = taskTrackingRepository.getInProgress()
+    if (inProgressTasks.isEmpty()) {
+      log.debug("No active orca tasks to monitor from ${javaClass.simpleName}.")
+    }
+
+    inProgressTasks
+      .forEach { taskId ->
+        val response = getTask(taskId)
+        if (response.status.isComplete()) {
+          val taskInfo = taskTrackingRepository.getTaskDetail(taskId)
+          if (taskInfo == null) {
+            log.error("Detail for task $taskId was not found in repository. Something went wrong!")
```

Comment:
This isn't a particularly helpful log message. If something went wrong, we should do what we can to provide good context into what actually went wrong. Think about 6 months, a year from now and we see this log message--will future Emily hate past Emily when triaging a production issue?

Reply:
Made it a bit better... 
`log.error("TaskDetail not found in tracking repository for $taskId. Unable to fire completion event for task.")` 
I don't know when this case would happen. Redis problems? The wrong tasks being cleaned out of the task repository? So I can't give a more helpful message.

Reply:
Hmm, based on the current implementation of the repository, a Redis problem would throw an exception and it wouldn't even get to this conditional. The updated log message is improved!